File: Core Graphics/CGContext/CGContext/DrawImageView.m
//
//  DrewImageView.m
//  CGContext
//
//  Created by pantosoft on 2018/3/7.
//  Copyright © 2018年 jianghongbing. All rights reserved.
//

#import "DrawImageView.h"

@implementation DrawImageView

- (void)drawRect:(CGRect)rect {
    NSString *imagePath = [[NSBundle mainBundle] pathForResource:@"Demo" ofType:@"png"];
    UIImage *image = [UIImage imageWithContentsOfFile:imagePath];
    CGContextRef ctx = UIGraphicsGetCurrentContext();
    //1.绘制图片,以填冲的形式来绘制
    CGContextDrawImage(ctx, CGRectMake(20, 80, image.size.width, image.size.height), image.CGImage);
    CGContextDrawImage(ctx, CGRectMake(120, 80, 200, 200), image.CGImage);
    
    CGRect tiledImageRect = CGRectMake(20, 300, rect.size.width - 40, rect.size.height - 300 - 10);
    //2.将绘制的区域固定在指定的区域,相当于裁剪到指定区域之外的区域
    CGContextClipToRect(ctx, tiledImageRect);
    //3.以平铺的形式来绘制图片,默认会平铺整个上下区域
    CGContextDrawTiledImage(ctx, CGRectMake(10, 10, image.size.width, image.size.height), image.CGImage);
    CGContextSetRGBFillColor(ctx, 1.0, 0, 0, 0.5);
    CGContextSetStrokeColorWithColor(ctx, [UIColor blueColor].CGColor);
    CGContextSetLineWidth(ctx, 3.0);
    CGContextFillRect(ctx, CGRectMake(10, 10, image.size.width, image.size.height));
    CGRect clipRect = CGContextGetClipBoundingBox(ctx);
    CGContextStrokeRect(ctx, clipRect);
    CGContextDrawPDFPage(ctx, NULL);
}
@end

File: UIKit/UIImage/UIImage/DrawImageView.swift
//
//  DrawImageView.swift
//  UIImage
//
//  Created by pantosoft on 2017/11/29.
//  Copyright © 2017年 jianghongbing. All rights reserved.
//

import UIKit

class DrawImageView: UIView {
    override func draw(_ rect: CGRect) {
        if let image = UIImage.init(named: "1") {
            //1.从某个点开始绘制图片
            //        image.draw(at: CGPoint(x: rect.midX, y: rect.midY))
            //2.在某个区域中绘制图片
//            image.draw(in: CGRect(x: 0, y: 0, width: image.size.width * 2, height: image.size.height * 2))
            //3.在某个区域中,以平铺的方式绘制图片
//            image.drawAsPattern(in: CGRect(x: 0, y: 0, width: image.size.width * 2, height: image.size.height * 2))
            //4.带有混合模式的绘制图片
            image.draw(at: .zero, blendMode: .darken, alpha: 0.5)
            image.draw(in:CGRect(x: 0, y: 0, width: image.size.width * 2, height: image.size.height * 2) , blendMode: .color, alpha: 0.8)
        }
    }
}
